forward/request_sender.go

```diff
@@ -208,13 +237,59 @@ func (s *requestSender) MakeCall(ctx context.Context, res *[]byte, fwdError *err
 			return
 		}
 
-		*res = arg3
 		done <- true
 	}()
 
 	return done
 }
 
+func forwardArgs(call *tchannel.OutboundCall, arg2, arg3 []byte, res *[]byte) error {
+	if err := writeArg(call.Arg2Writer, arg2); err != nil {
+		return err
+	}
+	if err := writeArg(call.Arg3Writer, arg3); err != nil {
+		return err
+	}
+
+	resp := call.Response()
+	if err := copyArg(ioutil.Discard, resp.Arg2Reader); err != nil {
+		return err
+	}
+
+	buf := bufPool.Get()
+	defer bufPool.Put(buf)
+	if err := copyArg(buf, resp.Arg3Reader); err != nil {
+		return err
+	}
+
+	*res = append(*res, buf.Bytes()...)
+	return nil
+}
+
+func writeArg(writable func() (tchannel.ArgWriter, error), bs []byte) error {
+	aw, err := writable()
+	if err != nil {
+		return err
+	}
+	_, err = aw.Write(bs)
+	if err == nil {
+		err = aw.Close()
+	}
+	return err
+}
+
+func copyArg(w io.Writer, readable func() (tchannel.ArgReader, error)) error {
+	ar, err := readable()
+	if err != nil {
+		return err
+	}
+	_, err = io.Copy(w, ar)
```

Comment:
this is going to allocate an internal buffer, is that ok?

Reply:
```
    If src implements the WriterTo interface, the copy is implemented by
    calling src.WriteTo(dst). Otherwise, if dst implements the ReaderFrom
    interface, the copy is implemented by calling dst.ReadFrom(src).
```

And both of our writers implement `io.ReaderFrom`, you taught me this when you asked for the `ioutil.Discard.ReadFrom` point ;-)

As a curiosity;
- I believe the order of preference expressed in `io.Copy` docs above...
- ...means that you should prefer to have a src that implements `io.WriterTo` when possible
- ...so I guess we could consider doing so in `tchannel-go.fragmentingReader` and/or the arg reader helper

This PR has just been a font of learning for me at least...

Reply:
Ah yes, since the call sites only pass in writers with `ReadFrom`, this will be efficient. To make sure things stay efficient, consider accepting an interface with `ReadFrom` (will need to cast `ioutil.Discard` which should succeed, can use a global to verify that on startup: `var _ io.ReaderFrom = ioutil.Discard.(io.ReaderFrom)`